### PR TITLE
Replace iodb with leveldb

### PR DESCRIPTION
**Pull request by knizhnik**

### CODE REVIEW

File: avldb/src/main/scala/scorex/db/LDBVersionedStore.scala [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+    iterator.close()
+    // As far as org.iq80.leveldb doesn't support iteration in reverse order, we have to iterate in the order
+    // of decreasing LSNs and then revert version list. For each version we store first (smallest) LSN.
+    versionLsn += lastLsn // first LSN of oldest version
+    versionLsn = versionLsn.reverse // LSNs should be in ascent order
+    versionLsn.remove(versionLsn.size - 1) // remove last element which corresponds to next assigned LSN
+    versions.reverse
+  }
+
+  case class Undo(versionID: VersionID, key: Array[Byte], value : Array[Byte])
```
Comment (aslesarenko) [inline avldb/src/main/scala/scorex/db/LDBVersionedStore.scala]: Please describe what this class represents (its role)

File: avldb/src/test/scala/scorex/crypto/authds/avltree/batch/benchmark/IODBBenchmark.scala [hunk 2]
```diff
@@ -11,7 +12,8 @@ object IODBBenchmark extends App with FileHelper {
   val NumMods = 2000000
   val Step = 1000
 
-  val store = new LSMStore(getRandomTempDir)
+  //val store = new LSMStore(getRandomTempDir)
```
Comment (aslesarenko) [inline avldb/src/test/scala/scorex/crypto/authds/avltree/batch/benchmark/IODBBenchmark.scala]:
```suggestion
```

File: avldb/src/test/scala/scorex/crypto/authds/avltree/batch/benchmark/OOMTest.scala [hunk 3]
```diff
@@ -20,7 +20,8 @@ object OOMTest extends App {
   protected implicit val hf = Blake2b256
 
   val dir: File = Files.createTempDirectory("oom-test").toFile
-  val store = new LSMStore(dir, keepVersions = 200)
+  //val store = new LSMStore(dir, keepVersions = 200)
```
Comment (aslesarenko) [inline avldb/src/test/scala/scorex/crypto/authds/avltree/batch/benchmark/OOMTest.scala]:
```suggestion
```

File: src/main/scala/org/ergoplatform/utils/FileUtils.scala [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package org.ergoplatform.utils
+
+import java.io.File
+
+object FileUtils {
+
+  def deleteRecursive(dir: File): Unit = {
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/utils/FileUtils.scala]: Every new public method require ScalaDoc

**Comment by aslesarenko** [comment]
@greenhat Could you please help to solve it-test failing problem.

**Comment by greenhat** [comment]
I can look at it tomorrow morning thoroughly but from the quick look at logs it seems like some dependency problem with logback.

**Comment by greenhat** [comment]
@aslesarenko @knizhnik I fixed sbt assembly in https://github.com/ergoplatform/ergo/pull/969/commits/12f544abd2525962f2b66c3246d1fc8871a4acb4 (`module-info.class` was moved to `META-INF/versions/9`);
see https://stackoverflow.com/questions/54834125/sbt-assembly-deduplicate-module-info-class#comment103351597_55557287

File: avldb/benchmarks/src/main/scala/scorex/crypto/authds/benchmarks/AVLTreeBatchPerformance.scala [hunk 5]
```diff
@@ -20,7 +21,8 @@ object AVLTreeBatchPerformance extends {
 
     val logger = LoggerFactory.getLogger("TEST")
     var prover: Prover = _
-    var store: LSMStore = _
+	//var store: LSMStore = _
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply avldb/benchmarks/src/main/scala/scorex/crypto/authds/benchmarks/AVLTreeBatchPerformance.scala]: Done

File: avldb/src/main/scala/scorex/db/LDBFactory.scala [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+import org.iq80.leveldb.{DB, Range, DBFactory, DBIterator, Options, ReadOptions, Snapshot, WriteBatch, WriteOptions}
+import scorex.util.ScorexLogging
+
+import scala.collection.mutable
+import scala.util.Try
+
+// Registry of opened LevelDB instances.
+// LevelDB prohibit access to the same storage file from more than one DB instance.
+// And ergo application (mostly tests) quit frequently doesn't not explicitly close
+// database and tries to reopen it.
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply avldb/src/main/scala/scorex/db/LDBFactory.scala]: Done

File: avldb/src/main/scala/scorex/db/LDBFactory.scala [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+case class StoreRegistry(val factory : DBFactory) extends DBFactory {
+
+  val lock = new ReentrantReadWriteLock()
+  val map = new mutable.HashMap[File, RegisteredDB]
+
+  // Decorator of LevelDB DB class which overrides close() methods and unlinks database from registry on close.
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply avldb/src/main/scala/scorex/db/LDBFactory.scala]: It was deteriorate code - removed

File: avldb/src/main/scala/scorex/db/LDBFactory.scala [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+  case class RegisteredDB(val impl:DB, val path: File) extends DB {
+    var count: Int = 0
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply avldb/src/main/scala/scorex/db/LDBFactory.scala]: Not used any more: removed

File: avldb/src/main/scala/scorex/db/LDBVersionedStore.scala [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+import io.iohk.iodb.{ByteArrayWrapper, Store}
+import org.iq80.leveldb._
+import java.nio.ByteBuffer
+
+import scala.collection.mutable.ArrayBuffer
+import java.util.concurrent.locks.ReentrantReadWriteLock
+
+
+//
+// Implementation of verioned storage on top of leveldb.
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply avldb/src/main/scala/scorex/db/LDBVersionedStore.scala]: Done

File: avldb/src/main/scala/scorex/db/LDBVersionedStore.scala (hunk 1 above)
Comment (knizhnik) [reply avldb/src/main/scala/scorex/db/LDBVersionedStore.scala]: Done

File: avldb/src/test/scala/scorex/crypto/authds/avltree/batch/benchmark/IODBBenchmark.scala (hunk 2 above)
Comment (knizhnik) [reply avldb/src/test/scala/scorex/crypto/authds/avltree/batch/benchmark/IODBBenchmark.scala]: Done

File: avldb/src/test/scala/scorex/crypto/authds/avltree/batch/benchmark/OOMTest.scala (hunk 3 above)
Comment (knizhnik) [reply avldb/src/test/scala/scorex/crypto/authds/avltree/batch/benchmark/OOMTest.scala]: Done

File: src/main/scala/org/ergoplatform/utils/FileUtils.scala (hunk 4 above)
Comment (knizhnik) [reply src/main/scala/org/ergoplatform/utils/FileUtils.scala]: Done

File: avldb/src/main/scala/scorex/db/LDBVersionedStore.scala [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+  }
+
+  private def newLSN(): Array[Byte] = {
+    lsn += 1
+    encodeLSN(lsn)
+  }
+
+  // Invert word to provide descending key order.
+  // Java implementation of LevelDB org.iq80.leveldb doesn't support iteration in backward direction.
+  private def decodeLSN(lsn: Array[Byte]): LSN = {
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply avldb/src/main/scala/scorex/db/LDBVersionedStore.scala]: Done

File: avldb/src/main/scala/scorex/db/LDBVersionedStore.scala [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+// This class implements IODB io.iohk.iodb.Store interface.
+// Only this interface is used from IODB to avoid changes in existed code which works with this interface.
+// LevelDB implementation of versioned store is based on maintaining "compensating transaction" list,
+// list of reverse operations needed to undo changes of applied transactions.
+// This list is stored in separate levedb database (undo) and size of list is limited by keepVersions parameter.
+// If keepVersions isd 0, then undo list is not maintained and rollback of the committed transactions is no possible.
+//
+class LDBVersionedStore(val dir: File, val keepVersions: Int = 0) extends Store {
+  type LSN = Long // logical serial number: type used to provide order of records in undo list
+  private val db: DB = createDB(dir, "ldb_main")
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply avldb/src/main/scala/scorex/db/LDBVersionedStore.scala]: Done

**Pull request merged**